File: Binary Search/Aggressive-Cows/Better/better.java
import java.util.*;

public class better {
    static boolean canPlace(int[] stalls, int cows, int dist) {
        int count = 1, last = stalls[0];
        for (int i = 1; i < stalls.length; i++) {
            if (stalls[i] - last >= dist) {
                count++;
                last = stalls[i];
            }
            if (count >= cows) return true;
        }
        return false;
    }

    static int aggressiveCows(int[] stalls, int cows) {
        Arrays.sort(stalls);
        int low = 1, high = stalls[stalls.length - 1] - stalls[0];
        int ans = 0;

        for (int dist = high; dist >= 1; dist--) {
            if (canPlace(stalls, cows, dist)) {
                ans = dist;
                break;
            }
        }
        return ans;
    }

    public static void main(String[] args) {
        int[] stalls = {1, 2, 8, 4, 9};
        int cows = 3;
        System.out.println(aggressiveCows(stalls, cows));
    }
}

/*
Time Complexity: O(N * (max(stall) - min(stall))) worst case,
but faster in practice because we stop at the first valid distance.
Space Complexity: O(1)
*/

File: Binary Search/Aggressive-Cows/Better/better.py
def can_place(stalls, cows, dist):
    count = 1
    last = stalls[0]
    for i in range(1, len(stalls)):
        if stalls[i] - last >= dist:
            count += 1
            last = stalls[i]
        if count >= cows:
            return True
    return False

def aggressive_cows(stalls, cows):
    stalls.sort()
    low, high = 1, stalls[-1] - stalls[0]
    ans = 0

    for dist in range(high, 0, -1):  # start from max possible distance
        if can_place(stalls, cows, dist):
            ans = dist
            break
    return ans

stalls = [1, 2, 8, 4, 9]
cows = 3
print(aggressive_cows(stalls, cows))

"""
Time Complexity: O(N * (max(stall) - min(stall))) worst case,
but faster in practice due to early stopping.
Space Complexity: O(1)
"""
